route-rs-runtime/src/link/async_link.rs

```diff
@@ -214,56 +269,105 @@ impl<Packet: Sized> Stream for AsyncEgressor<Packet> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::element::{AsyncIdentityElement, DropElement, IdentityElement, TransformElement};
+    use crate::element::{DropElement, IdentityElement, TransformElement};
     use crate::link::sync_link::SyncLinkBuilder;
-    use crate::link::{ElementLinkBuilder, LinkBuilder};
+    use crate::link::{ElementLinkBuilder, LinkBuilder, TokioRunnable};
     use crate::utils::test::packet_collectors::ExhaustiveCollector;
     use crate::utils::test::packet_generators::{immediate_stream, PacketIntervalGenerator};
     use core::time;
     use futures::future::lazy;
 
+    fn run_tokio(runnables: Vec<TokioRunnable>) {
+        tokio::run(lazy(|| {
+            for runnable in runnables {
+                tokio::spawn(runnable);
+            }
+            Ok(())
+        }));
+    }
+
+    #[test]
+    #[should_panic]
+    fn panics_when_built_without_input_streams() {
+        let identity_element: IdentityElement<i32> = IdentityElement::new();
+
+        AsyncLinkBuilder::new()
+            .element(identity_element)
+            .build_link();
+    }
+
+    #[test]
+    #[should_panic]
+    fn panics_when_built_without_element() {
+        let packets: Vec<i32> = vec![];
+        let packet_generator: PacketStream<i32> = immediate_stream(packets.clone());
+
+        AsyncLinkBuilder::<IdentityElement<i32>>::new()
+            .ingressor(packet_generator)
+            .build_link();
+    }
+
+    #[test]
+    fn builder_methods_work_in_any_order() {
+        let packets: Vec<i32> = vec![];
+
+        let packet_generator0 = immediate_stream(packets.clone());
+        let identity_element0 = IdentityElement::new();
+
+        AsyncLinkBuilder::new()
+            .ingressor(packet_generator0)
+            .element(identity_element0)
+            .build_link();
+
+        let packet_generator1 = immediate_stream(packets.clone());
+        let identity_element1 = IdentityElement::new();
+
+        AsyncLinkBuilder::new()
+            .element(identity_element1)
+            .ingressor(packet_generator1)
+            .build_link();
+    }
+
     #[test]
     fn async_link() {
-        let default_channel_size = 10;
         let packets = vec![0, 1, 2, 420, 1337, 3, 4, 5, 6, 7, 8, 9];
         let packet_generator = immediate_stream(packets.clone());
 
-        let elem = AsyncIdentityElement::new();
+        let elem = IdentityElement::new();
 
-        let link = AsyncLink::new(Box::new(packet_generator), elem, default_channel_size);
+        let (mut runnables, mut egressors) = AsyncLinkBuilder::new()
+            .ingressors(vec![Box::new(packet_generator)])
```

Comment:
For each of these test examples, you can use the `ingressor` function you defined. `immediate_stream` is actually already inside a `Box`, so you can just  `.ingressor(packet_generator)`

Reply:
I'm not sure that we should have an `ingressor` function at all. I think it kind of breaks our own trait concept? In a similar vein, for other functions, like braches, queue_length, etc, should we pull those into the trait and then make the links that don't need them implement them as no-ops?

Reply:
I see it as a convenience method that isn't part of the `Link` trait. Something that users can specify when creating their own HoL (higher-order links) or composites, but obviously something we can't do in graphgen. As long as we're sure that

```builder.ingressor(in) == builder.ingressors(vec![Box::new(in)])```

then nothing is broken, and users get a slighly more ergonomic method to use.

Reply:
Hmm, put that way,  I do like having it there.  Might form a nice marker for the user to know that the link is 1 to stuff.

Reply:
Went ahead and implemented this for all links. On links that take multiple inputs, it works like a push() function. One links that only take one input, works as replace()